custom_components/localtuya/light.py

```diff
@@ -141,31 +167,66 @@ async def async_turn_on(self, **kwargs):
                 self._upper_brightness,
             )
             if self._is_white_mode:
-                await self._device.set_dp(brightness, self._config.get(CONF_BRIGHTNESS))
+                states[self._config.get(CONF_BRIGHTNESS)] = brightness
+
             else:
-                color = "{:04x}{:04x}{:04x}".format(
-                    round(self._hs[0]), round(self._hs[1] * 10.0), brightness
-                )
-                await self._device.set_dp(color, self._config.get(CONF_COLOR))
+                if self._is_rgb_mode:
+                    rgb = color_util.color_hsv_to_RGB(
+                        self._hs[0],
+                        self._hs[1],
+                        int(brightness * 100 / self._upper_brightness),
+                    )
+                    color = "{:02x}{:02x}{:02x}{:04x}{:02x}{:02x}".format(
+                        round(rgb[0]),
+                        round(rgb[1]),
+                        round(rgb[2]),
+                        round(self._hs[0]),
+                        round(self._hs[1] * 255 / 100),
+                        brightness,
+                    )
+                else:
+                    color = "{:04x}{:04x}{:04x}".format(
+                        round(self._hs[0]), round(self._hs[1] * 10.0), brightness
+                    )
+                states[self._config.get(CONF_COLOR)] = color
 
         if ATTR_HS_COLOR in kwargs and (features & SUPPORT_COLOR):
             hs = kwargs[ATTR_HS_COLOR]
-            color = "{:04x}{:04x}{:04x}".format(
-                round(hs[0]), round(hs[1] * 10.0), self._brightness
-            )
-            await self._device.set_dp(color, self._config.get(CONF_COLOR))
+            if self._is_rgb_mode is None:
+                self._is_rgb_mode = len(self.dps_conf(CONF_COLOR)) > 12
+
+            if self._is_rgb_mode:
```

Comment:
Maybe we can extract this part to a method and re-use it in `async_turn_on` as well?

Reply:
Eventually, i extracted the check in a method without using variable.